### PR TITLE
CScreenSaverUtils: add class to manage screensaver and dpms

### PR DESCRIPTION
**Pull request by lrusak**
This breaks out the screensaver code from `CApplication` and makes it listen on it's own `CApplicationMessenger` bus. This seemed logical as it `TMSG_ACTIVATESCREENSAVER` was on the bus already.

I'm submitting this early as I'm going to be away for a few weeks. It's still in pretty rough shape but it works for me (tm) with kodi-gbm.

Tested and needs testing:
 - [x] black and dim screensavers
 - [x] python screensavers
 - [x] binary screensavers
 - [ ] OS screen saver inhibiting.
 - [ ] extensive screensaver and dpms activation testing

Todo:
 - [ ]  add lockscreen locking after screensaver deactivation.

I had to add an extra timer to debounce any input so the screensaver doesn't deactivate right away. If there is a better solution to this I'm all ears.

I've also updated `CGUIWindowScreensaverDim` to what I think is more logical.

### CODE REVIEW

File: xbmc/utils/ScreenSaverUtils.cpp [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+  if (!settings)
+    return;
+
+  std::string screensaver = settings->GetString(CSettings::SETTING_SCREENSAVER_MODE);
+
+  if (screensaver == "screensaver.xbmc.builtin.dim" ||
+      screensaver == "screensaver.xbmc.builtin.black" ||
+      screensaver.empty())
+  {
+    CServiceBroker::GetGUI()->GetWindowManager().CloseDialogs();
```
Comment (FernetMenta) [inline xbmc/utils/ScreenSaverUtils.cpp]: that's really bad design. GUI methods must only be called by the render thread. I think this is still the case here but far from obvious. One the one hand you use AppMessanger for sending messages from main thread to main thread, on other hand you call from some util class directly into GUI component.

File: xbmc/utils/ScreenSaverUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+  if (!settings)
+    return;
+
+  if (g_application.GetAppPlayer().IsPlayingAudio() && settings->GetBool(CSettings::SETTING_SCREENSAVER_USEMUSICVISINSTEAD))
+  {
+    std::string visualisation = settings->GetString(CSettings::SETTING_MUSICPLAYER_VISUALISATION);
+
+    if (!visualisation.empty())
+    {
+      CServiceBroker::GetGUI()->GetWindowManager().ActivateWindow(WINDOW_VISUALISATION);
```
Comment (FernetMenta) [inline xbmc/utils/ScreenSaverUtils.cpp]: must only be called by render thread but Activate is called by Process()

File: xbmc/utils/ScreenSaverUtils.cpp (hunk 1 above)
Comment (lrusak) [reply xbmc/utils/ScreenSaverUtils.cpp]: Is there another mechanism that can be used? What is the restriction about calling the GUI from the main thread?

File: xbmc/ServiceManager.h [hunk 3]
```diff
@@ -63,6 +63,11 @@ namespace PERIPHERALS
   class CPeripherals;
 }
 
+namespace UTILS
+{
+  class CScreenSaverUtils;
```
Comment (Rechi) [inline xbmc/ServiceManager.h]: do not indent namespaces

File: xbmc/interfaces/json-rpc/InputOperations.cpp [hunk 4]
```diff
@@ -20,21 +20,9 @@
...
-  if (g_application.WakeUpScreenSaverAndDPMS())
-    return true;
-
-  return false;
-}
-
 JSONRPC_STATUS CInputOperations::SendAction(int actionID, bool wakeScreensaver /* = true */, bool waitResult /* = false */)
 {
-  if(!wakeScreensaver || !handleScreenSaver())
+  if(!wakeScreensaver)
```
Comment (Rechi) [inline xbmc/interfaces/json-rpc/InputOperations.cpp]: space missing between `if` and `(`

File: xbmc/powermanagement/DPMSSupport.cpp [hunk 5]
```diff
@@ -8,6 +8,8 @@
 
 #include "DPMSSupport.h"
 
+#include "Application.h"
+#include "interfaces/AnnouncementManager.h"
```
Comment (Rechi) [inline xbmc/powermanagement/DPMSSupport.cpp]: move this one line below

File: xbmc/powermanagement/DPMSSupport.cpp [hunk 6]
```diff
@@ -42,3 +44,26 @@ bool CDPMSSupport::IsModeSupported(PowerSavingMode mode) const
...
+{
+  if (m_active)
+    return;
+
+  CServiceBroker::GetAnnouncementManager()->Announce(ANNOUNCEMENT::GUI, "xbmc", "OnDPMSActivated");
+  EnablePowerSaving(GetSupportedModes().front());
+  g_application.SetRenderGUI(false);
+  m_active = true;
+}
+
```
Comment (Rechi) [inline xbmc/powermanagement/DPMSSupport.cpp]: no need for an additional empty line at the end of file

File: xbmc/utils/ScreenSaverUtils.cpp [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "ScreenSaverUtils.h"
+
+#include "addons/AddonInfo.h"
+#include "addons/AddonManager.h"
+#include "addons/settings/GUIDialogAddonSettings.h"
+#include "Application.h"
```
Comment (Rechi) [inline xbmc/utils/ScreenSaverUtils.cpp]: move to top of this block

File: xbmc/utils/ScreenSaverUtils.cpp [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+#include "guilib/WindowIDs.h"
+#include "input/actions/Action.h"
+#include "input/actions/ActionIDs.h"
+#include "interfaces/AnnouncementManager.h"
+#include "interfaces/generic/ScriptInvocationManager.h"
+#include "messaging/ApplicationMessenger.h"
+#include "powermanagement/DPMSSupport.h"
+#include "pvr/PVRGUIActions.h"
+#include "pvr/PVRManager.h"
+#include "ServiceBroker.h"
```
Comment (Rechi) [inline xbmc/utils/ScreenSaverUtils.cpp]: move to top of this block (after Application.h)

File: xbmc/utils/ScreenSaverUtils.cpp [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+
+  CServiceBroker::GetGUI()->GetWindowManager().ActivateWindow(WINDOW_SCREENSAVER);
+}
+
+void CScreenSaverUtils::OnSettingChanged(std::shared_ptr<const CSetting> setting)
+{
+  if (!setting)
+    return;
+
+  const std::string &settingId = setting->GetId();
```
Comment (Rechi) [inline xbmc/utils/ScreenSaverUtils.cpp]: left align `&`

File: xbmc/utils/ScreenSaverUtils.cpp [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+    }
+  }
+}
+
+void CScreenSaverUtils::OnSettingAction(std::shared_ptr<const CSetting> setting)
+{
+  if (!setting)
+    return;
+
+  const std::string &settingId = setting->GetId();
```
Comment (Rechi) [inline xbmc/utils/ScreenSaverUtils.cpp]: left align `&`

File: xbmc/utils/ScreenSaverUtils.cpp [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+int CScreenSaverUtils::GetMessageMask()
+{
+  return TMSG_MASK_SCREENSAVER;
+}
+
+void CScreenSaverUtils::OnApplicationMessage(ThreadMessage* pMsg)
+{
+  uint32_t msg = pMsg->dwMessage;
+
+  switch(msg)
```
Comment (Rechi) [inline xbmc/utils/ScreenSaverUtils.cpp]: space missing between `switch` and `(`

File: xbmc/windows/GUIWindowScreensaverDim.cpp [hunk 12]
```diff
@@ -8,10 +8,13 @@
 
 #include "GUIWindowScreensaverDim.h"
 
+#include "addons/AddonManager.h"
```
Comment (Rechi) [inline xbmc/windows/GUIWindowScreensaverDim.cpp]: move after ServiceBroker.h

File: xbmc/windows/GUIWindowScreensaverDim.cpp [hunk 13]
```diff
@@ -69,3 +44,44 @@ void CGUIWindowScreensaverDim::Render()
...
+    else
+      m_newDimLevel = 100.0f;
+
+    break;
+  }
+  }
+
+  return CGUIWindow::OnMessage(message);
+}
+
```
Comment (Rechi) [inline xbmc/windows/GUIWindowScreensaverDim.cpp]: no need for an extra empty line at the end of file

**Comment by arnova** [comment]
@lrusak : What's the planning for this? It's preventing #16345 from being merged. If this stays unclear I just want to merge #16345 so we can at least move forward with that.

**Comment by lrusak** [comment]
@arnova go ahead with #16345

**Pull request closed without merging**